Multiparty-Constraint-Layout-Kotlin/app/src/main/java/com/tokbox/sample/multipartyconstraintlayout/MainActivity.kt

```diff
@@ -0,0 +1,289 @@
+    }
+
+    override fun onPermissionsDenied(requestCode: Int, perms: List<String>) {
+        finishWithMessage("onPermissionsDenied: $requestCode: $perms")
+    }
+
+    private fun getResIdForSubscriberIndex(index: Int): Int {
+        val arr = resources.obtainTypedArray(R.array.subscriber_view_ids)
+        val subId = arr.getResourceId(index, 0)
+        arr.recycle()
+        return subId
+    }
+
+    private fun startPublisherPreview() {
+        publisher = Publisher.Builder(this).build()
+        publisher?.setPublisherListener(publisherListener)
+        publisher?.setStyle(BaseVideoRenderer.STYLE_VIDEO_SCALE, BaseVideoRenderer.STYLE_VIDEO_FILL)
+        publisher?.startPreview()
+    }
+
+    @AfterPermissionGranted(PERMISSIONS_REQUEST_CODE)
+    private fun requestPermissions() {
+        val perms = arrayOf(Manifest.permission.INTERNET, Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO)
+
+        if (EasyPermissions.hasPermissions(this, *perms)) {
+            session = Session.Builder(this, OpenTokConfig.API_KEY, OpenTokConfig.SESSION_ID)
+                .sessionOptions(object : SessionOptions() {
+                    override fun useTextureViews(): Boolean {
+                        return true
+                    }
+                }).build()
+
+            session?.setSessionListener(sessionListener)
+            session?.connect(OpenTokConfig.TOKEN)
+            startPublisherPreview()
+            publisher?.view?.id = R.id.publisher_view_id
+            container.addView(publisher?.view)
+            calculateLayout()
+        } else {
+            EasyPermissions.requestPermissions(
+                this,
+                getString(R.string.rationale_video_app),
+                PERMISSIONS_REQUEST_CODE,
+                *perms
+            )
+        }
+    }
+
+    private fun calculateLayout() {
+        val set = ConstraintSetHelper(R.id.main_container)
+        val size = subscribers.size
+        if (size == 0) {
+            // Publisher full screen
+            set.layoutViewFullScreen(R.id.publisher_view_id)
+        } else if (size == 1) {
+            // Publisher
+            // Subscriber
+            set.layoutViewAboveView(R.id.publisher_view_id, getResIdForSubscriberIndex(0))
+            set.layoutViewWithTopBound(R.id.publisher_view_id, R.id.main_container)
+            set.layoutViewWithBottomBound(getResIdForSubscriberIndex(0), R.id.main_container)
+            set.layoutViewAllContainerWide(R.id.publisher_view_id, R.id.main_container)
+            set.layoutViewAllContainerWide(getResIdForSubscriberIndex(0), R.id.main_container)
+            set.layoutViewHeightPercent(R.id.publisher_view_id, .5f)
+            set.layoutViewHeightPercent(getResIdForSubscriberIndex(0), .5f)
+        } else if (size > 1 && size % 2 == 0) {
+            //  Publisher
+            // Sub1 | Sub2
+            // Sub3 | Sub4
+            //    .....
```

Comment:
Cognitive Complexity is more for this function.

Since this is an open source example, Maybe we need to make it more clear.
size > 1 && size % 2 == 0  -> MoreThanOneEvenAudience

Reply:
Agreed. But this is the same logic as in the Java example. So, maybe we can leave it as is for now.

Reply:
Agreed.